base/compiler/ssair/inlining.jl

```diff
@@ -445,6 +447,14 @@ function ir_inline_item!(compact::IncrementalCompact, idx::Int, argexprs::Vector
                 stmt′ = PhiNode(Int32[edge+bb_offset for edge in stmt′.edges], stmt′.values)
             end
             inline_compact[idx′] = stmt′
+            if extra_flags != 0 && !isa(stmt′, Union{GotoNode, GotoIfNot})
+                if (extra_flags & IR_FLAG_NOTHROW) != 0 && inline_compact[SSAValue(idx′)][:type] === Union{}
```

Comment:
I think we may need this sort of tweak for the other effects, e.g. `:consistent`-cy can now be proved in a presence of (unreturned) mutable allocations but it sounds quite wrong to mark the mutable allocation as `:consistent`.
Having said that, this may be enough for the meanwhile since we only look for `nothrow` and `effect_free` in our optimization passes, and I'm fine with leaving these handling for the future when the potential problem comes up.